ec/src/models/mod.rs

```diff
@@ -64,21 +66,32 @@ pub trait SWModelParameters: ModelParameters {
 
     /// Check if the provided curve point is in the prime-order subgroup.
     ///
-    /// The default implementation multiplies `item` by the order `r` of the prime-order
-    /// subgroup, and checks if the result is one.
+    /// The default implementation multiplies `item` by the order `r` of the
+    /// prime-order subgroup, and checks if the result is one.
     /// Implementors can choose to override this default impl
     /// if the given curve has faster methods
-    /// for performing this check (for example, via leveraging curve isomorphisms).
+    /// for performing this check (for example, via leveraging curve
+    /// isomorphisms).
     fn is_in_correct_subgroup_assuming_on_curve(
         item: &short_weierstrass_jacobian::GroupAffine<Self>,
     ) -> bool {
         item.mul_bits(BitIteratorBE::new(Self::ScalarField::characteristic()))
             .is_zero()
     }
+
+    /// Default implementation of group multiplication for projective
+    /// coordinates
+    fn mul(
+        base: &short_weierstrass_jacobian::GroupProjective<Self>,
+        scalar: &Self::ScalarField,
+    ) -> short_weierstrass_jacobian::GroupProjective<Self> {
+        base.mul(scalar.into_bigint())
```

Comment:
What I had in mind was actually something even more aggressive: to move the actual `mul` code here, and call `P::mul` from `GroupProjective::mul`. That is, the impl of this method in `GroupProjective` changes to
```rust
fn mul(...) {
    P::mul(...)
}
```

Also, we might want to add both `mul_projective` and `mul_affine` methods here.

Reply:
I just tried your suggestion.

- we cannot add it directly to `ModelParameter`: we need to specify if it is TE or SW
- we also cannot directly invoke it from `AffineCurve` or `ProjectiveCurve`, the trait is bounded by `ModelParameter` that does not specify `TE` or `SW`

working on the rest of the changes now

Reply:
a workaround is to add an associated type for `ModelParameters` but it will be ugly and is a breaking change

Reply:
Ah sorry, should have been clearer, I meant that in the current strategy, instead of calling `base.mul(...)` inside `SWModelParameters`, we should change the implementation as follows:
```rust
// inside SW/TEModelParameters:
pub trait SW/TEModelParameters {
	fn mul_projective(g: GroupProjective<Self>, scalar: ...) -> GroupProjective<Self> {
	    // default double-and-add impl
	} 
}

// inside `ProjectiveCurve`:
pub trait ProjectiveCurve {
	fn mul(...); // <- no default impl anymore
}

// inside sw/te::GroupProjective:
impl ProjectiveCurve for GroupProjective {
    ...
	
	fn mul() {
		P::mul_projective(...)
	}
}
```

Reply:
Got you. Should be fixed now.